extensions/warp-ipfs/src/store/community.rs

```diff
@@ -0,0 +1,389 @@
+use super::{conversation::message::MessageDocument, topics::ConversationTopic, PeerIdExt};
+use crate::store::DidExt;
+use chrono::{DateTime, Utc};
+use core::hash::Hash;
+use indexmap::{IndexMap, IndexSet};
+use ipld_core::cid::Cid;
+use rust_ipfs::{Ipfs, Keypair};
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+use warp::{
+    crypto::DID,
+    error::Error,
+    raygun::community::{
+        Community, CommunityChannel, CommunityChannelPermission, CommunityChannelPermissions,
+        CommunityChannelType, CommunityInvite, CommunityPermission, CommunityPermissions,
+        CommunityRole, RoleId,
+    },
+};
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CommunityRoleDocument {
+    pub id: RoleId,
+    pub name: String,
+    pub members: IndexSet<DID>,
+}
+impl CommunityRoleDocument {
+    pub fn new(name: String) -> Self {
+        Self {
+            id: Uuid::new_v4(),
+            name,
+            members: IndexSet::new(),
+        }
+    }
+}
+impl From<CommunityRoleDocument> for CommunityRole {
+    fn from(value: CommunityRoleDocument) -> Self {
+        let mut role = CommunityRole::default();
+        role.set_id(value.id);
+        role.set_name(value.name);
+        role.set_members(value.members);
+        role
+    }
+}
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct CommunityInviteDocument {
+    pub id: Uuid,
+    pub target_user: Option<DID>,
+    pub created: DateTime<Utc>,
+    pub expiry: Option<DateTime<Utc>>,
+}
```

Comment:
I wonder if this should be signed. Eg signed by either the creator (or owner) or by the one who has (or had) the role or permission to send the invite (which could also be logged into the document to cross reference it). Thoughts?

Reply:
True, not a bad idea. Thinking more generally though, (not just about invitations), I think it might be better if we took an approach where we signed the inputs of change actions, rather than the resulting data. 

Reason is, the resulting data could be different depending on the order actions occur. But if we sign the inputs instead, that accurately reflects what the change action is really intended to be. 

And we could then also solve ordering inconsistency by caching the signed change inputs (recent ones) which would allow us to reorder and replay the changes for conflict resolution in case of a desync. 

Anyhow, I think I'd rather address this in another PR.